### PR TITLE
Merge Lukas start2.0 - Add scaling for the gamePanel

### LINKED ISSUE #48: Fix window scaling
source: https://www.codejava.net/java-se/graphics/drawing-an-image-with-automatic-scaling

### LINKED ISSUE #58: Finish off scaling
What's left: 
- Change the dice roll indicator from a Swing element to a g2d drawnString.
- Scale Swing elements

_Originally posted by @Tacitor in https://github.com/Evan-Kreutzwiser/ICS4U-Settlers-Of-Catan/issues/48#issuecomment-730858500_

### PR DESCRIPTION
**Pull request by Tacitor**
I will leave comments explaining what I have done in the code. But basically, I have added locked ratio scaling to all (I hope) aspects of the gamePanel 

resolves #48 
resolves #58

### CODE REVIEW

File: KrampitzKreutzwiserSettlersOfCatan/src/dataFiles/OldCode.java [hunk 1]
```diff
@@ -1,108 +1,128 @@
...
-        g2d.drawImage(tiles.get(18).getImage(), 1920 / 2 - (150 / 2) + (112 * 2), 1080 / 2 - (130 / 2) + (130), null);
-        */
-    
-}
+/*
+ * Lukas Krampitz
+ * Nov 8, 2020
+ * Just a file with some old code that we should keep just in case. This will never run from here
+ */
+package dataFiles;
```
Comment (Tacitor) [inline KrampitzKreutzwiserSettlersOfCatan/src/dataFiles/OldCode.java:6]: GH seems to think I basically rewrote the whole comment here but I am pretty sure I did not

File: KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GameFrame.java [hunk 2]
```diff
@@ -25,7 +31,14 @@ public GameFrame(MainMenu m) {
      */
     private void initFrame() {
         setTitle("Settlers of Catan");
-        setSize(1920, 1080); //set the size to 1080p
+        setSize(1920, 1080); //set the size to 1080p as a back up      
+        //setSize(1080, 1920);
+        //setSize(1280, 720);
```
Comment (Tacitor) [inline KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GameFrame.java:36]: These are just testing resolutions I found them useful to make sure everything works okay

File: KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.form [hunk 3]
```diff
@@ -34,17 +34,9 @@
           <Group type="102" attributes="0">
               <EmptySpace max="-2" attributes="0"/>
               <Group type="103" groupAlignment="0" attributes="0">
```
Comment (Tacitor) [inline KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.form:36]: In this file, I just removed a couple of Swing elements that I replaced with drawnString from g2d. Just make it easier to scale. Everything should still work the same way though.

File: KrampitzKreutzwiserSettlersOfCatan/src/textures/ImageRef.java [hunk 4]
```diff
@@ -5,10 +5,18 @@
  */
 package textures;
 
+import java.awt.Image;
+import javax.swing.ImageIcon;
+
```
Comment (Tacitor) [inline KrampitzKreutzwiserSettlersOfCatan/src/textures/ImageRef.java:10]: Prepared this mainly to hold all the Image objects but ran into some issues after the merge. I'll get back to this later.

File: KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java [hunk 5]
```diff
@@ -25,6 +25,7 @@
 import javax.swing.JFileChooser;
 import javax.swing.JOptionPane;
 import textures.ImageRef;
+import static textures.ImageRef.*;
```
Comment (Tacitor) [inline KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java:28]: Currently not being used. See ImageRef.java for details

File: KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java [hunk 6]
```diff
@@ -176,6 +192,65 @@ public final void mouseClicked(MouseEvent evt) {
 
         // Set the state of the builds buttons for the first player
         updateBuildButtons();
+
```
Comment (Tacitor) [inline KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java:195]: Initialize all the stuff used in scaling

File: KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java [hunk 7]
```diff
@@ -226,20 +301,23 @@ public void actionPerformed(java.awt.event.ActionEvent evt) {
         instructionLbl.setFont(new java.awt.Font("Times New Roman", 0, 18)); // NOI18N
         instructionLbl.setText("Place two roads and two small settlements each to start.");
 
-        buildMenuLbl.setFont(new java.awt.Font("Times New Roman", 0, 14)); // NOI18N
+        buildMenuLbl.setFont(new java.awt.Font("Times New Roman", 0, 18)); // NOI18N
```
Comment (Tacitor) [inline KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java:303]: Just changing up the font sizes here

File: KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java [hunk 8]
```diff
@@ -250,11 +328,15 @@ public void actionPerformed(java.awt.event.ActionEvent evt) {
...
+        backNoSaveBtn.addActionListener(new java.awt.event.ActionListener() {
+            public void actionPerformed(java.awt.event.ActionEvent evt) {
+                backNoSaveBtnActionPerformed(evt);
+            }
+        });
 
-        diceRollPromptLbl1.setFont(new java.awt.Font("Times New Roman", 0, 18)); // NOI18N
-        diceRollPromptLbl1.setText("You rolled a: ");
+        titleLbl.setFont(new java.awt.Font("Times New Roman", 1, 40)); // NOI18N
+        titleLbl.setText("Settlers of Catan");
```
Comment (Tacitor) [inline KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java:338]: GH is thinking I changed these lines. They are in fact NOT related and for 2 completely different Swing variables.

File: KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java [hunk 9]
```diff
@@ -263,16 +345,9 @@ public void actionPerformed(java.awt.event.ActionEvent evt) {
             .addGroup(layout.createSequentialGroup()
                 .addContainerGap()
                 .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
-                    .addGroup(javax.swing.GroupLayout.Alignment.TRAILING, layout.createSequentialGroup()
```
Comment (Tacitor) [inline KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java:266]: The dice roll indicator I deleted

File: KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java [hunk 10]
```diff
@@ -767,29 +863,29 @@ public boolean save() {
      * @param writeAdress
      * @throws FileNotFoundException
      */
-    public boolean writeToFile(String writeAdress) throws FileNotFoundException {
+    private boolean writeToFile(String writeAdress) throws FileNotFoundException {
```
Comment (Tacitor) [inline KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java:869]: Thought some of these should be private

File: KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java [hunk 11]
```diff
@@ -1153,7 +1256,7 @@ private void diceRoll() {
         roll += (int) (Math.random() * 6) + 1;
 
         // Display the number rolled to the user
-        diceRollLbl.setText(Integer.toString(roll));
+        diceRollVal = (Integer.toString(roll));
```
Comment (Tacitor) [inline KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java:1262]: drawn with g2d now

File: KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java [hunk 12]
```diff
@@ -1453,7 +1595,7 @@ else if (settlement.isLarge() == false) { // Small settlement
             // Get the number of cards the player has
             int listSize = cards[currentPlayer].size();
             // Calculate where the first card must go to center the list
-            int startPosition = 960 - (listSize * CARD_CLAY.getWidth(null) + (listSize - 1) * 10) / 2;
+            int startPosition = (int) ((superFrame.getWidth() / 2) - (listSize * getImgWidth(CARD_CLAY) + (listSize - 1) * (10 / scaleFactor)) / 2);
```
Comment (Tacitor) [inline KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java:1601]: I hope I did this right @Evan-Kreutzwiser

File: KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java [hunk 13]
```diff
@@ -1474,17 +1616,77 @@ else if (settlement.isLarge() == false) { // Small settlement
                     case 4: // Sheep card
                         image = CARD_SHEEP;
                         break;
-                    default: // 5: Ore card
+                    case 5: // 5: Ore card
                         image = CARD_ORE;
                         break;
+                    default: //error "card"
```
Comment (Tacitor) [inline KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java:1625]: Any other value would have just been an extra ore card so I added another "error" card

File: KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java [hunk 14]
```diff
@@ -1474,17 +1616,77 @@ else if (settlement.isLarge() == false) { // Small settlement
...
+    }
+    
+    /**
+     * Calculates the new scaled width of an image with a locked aspect ratio.
+     * @param image
+     * @return 
+     */
+    public int getImgWidth(Image image) {
+
+        if (superFrame.getWidth() > superFrame.getHeight()) {
```
Comment (Tacitor) [inline KrampitzKreutzwiserSettlersOfCatan/src/krampitzkreutzwisersettlersofcatan/GamePanel.java:1674]: Let me know if you think this needs explaining

**Review by Tacitor — COMMENTED** [review]
I think It's ready for merge.

**Review by Evan-Kreutzwiser — APPROVED** [review]
Looks good, and now probably the most challenging feature is done!

**Pull request merged**